x-pack/plugins/security/server/lib/authorization/has_privileges.test.js

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { hasPrivilegesWithServer, HAS_PRIVILEGES_RESULT } from './has_privileges';
+import { getClient } from '../../../../../server/lib/get_client_shield';
+import { DEFAULT_RESOURCE } from '../../../common/constants';
+import { getLoginPrivilege, getVersionPrivilege } from '../privileges';
+
+jest.mock('../../../../../server/lib/get_client_shield', () => ({
+  getClient: jest.fn()
+}));
+
+const defaultVersion = 'default-version';
+const defaultApplication = 'default-application';
+const defaultKibanaIndex = 'default-index';
+const savedObjectTypes = ['foo-type', 'bar-type'];
+
+const createMockServer = ({ settings = {} } = {}) => {
+  const mockServer = {
+    config: jest.fn().mockReturnValue({
+      get: jest.fn()
+    }),
+    log: jest.fn()
+  };
+
+  const defaultSettings = {
+    'pkg.version': defaultVersion,
+    'xpack.security.rbac.application': defaultApplication,
+    'kibana.index': defaultKibanaIndex,
+  };
+
+  mockServer.config().get.mockImplementation(key => {
+    return key in settings ? settings[key] : defaultSettings[key];
+  });
+
+  mockServer.savedObjects = {
+    types: savedObjectTypes
+  };
+
+  return mockServer;
+};
+
+const mockApplicationPrivilegeResponse = ({ hasAllRequested, privileges, application = defaultApplication, username = '' }) =>{
+  return {
+    username: username,
+    has_all_requested: hasAllRequested,
+    application: {
+      [application]: {
+        [DEFAULT_RESOURCE]: privileges
+      }
+    }
+  };
+};
+
+const mockKibanaIndexPrivilegesResponse = ({ privileges, index = defaultKibanaIndex }) => {
+  return {
+    index: {
+      [index]: privileges
+    }
+  };
+};
+
+const createMockCallWithRequest = (responses) => {
+  const mockCallWithRequest = jest.fn();
+  getClient.mockReturnValue({
+    callWithRequest: mockCallWithRequest
+  });
+
+  for (const response of responses) {
+    mockCallWithRequest.mockImplementationOnce(async () => response);
+  }
+
+  return mockCallWithRequest;
+};
+
+const expectNoDeprecationLogged = (mockServer) => {
+  expect(mockServer.log).not.toHaveBeenCalled();
```

Comment:
This should really be testing more than that the log function has been called, as you can call the log function for any number of purposes that aren't all for deprecation logging. We should at least check that deprecation is an argument and perhaps even snapshot test the message.